Use `Link` in `ArtemisClient`

klavs opened the pull request:
I've implemented the concept of a `Link` in `package:gql/link`.

Using `Link` in `ArtemisClient` would unlock the possibility to do request deduplication, request batching, request retries, etc.

All of these features could be implemented externally and shared between different client implementations which would benefit every client.

A side effect of this change is that GraphQL documents must be parsed before passing them to `Link`. Currently, it is done at run-time, but it would be easy to do it at build-time using `package:gql/dart` (see [this example](https://github.com/gql-dart/gql/tree/master/gql/example/ast_builder) and [this example](https://github.com/gql-dart/gql/blob/master/gql/lib/ast_builder.dart)).

comigor commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
 
-  _DefaultHttpJsonClient();
+  /// Construct the Link
+  SimpleHttpJsonLink(this.graphQLEndpoint);
```
As we'll already have `SimpleHttpJsonLink` to use as a default for `ArtemisClient`'s link, I think it should optionally receive a `http.Client` on its constructor (so it's easier to inject credentials and tokens, for example).

klavs replied:
That makes sense for compatibility reasons.

leoiacovini replied:
Agree that receiving the client is an important feature (or we would need to write a custom Link just to be able to wrap our client)

comigor commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
-  http.Client httpClient;
+  /// Instantiates an [ArtemisClient].
+  ArtemisClient({
+    this.link,
```
I think `link` could be optional (defaulting to `SimpleHttpJsonLink`, but receiving and forwarding `graphQLEndpoint` again) so final user won't have to initialize two variables to use this and this PR won't be a breaking change.

klavs replied:
If breaking changes is a consern, we can create a new factory constructor.

But I think a longer term solution is to make `ArtemisClient` protocol-agnostic (as is the GraphQL spec) so using `SimpleHttpJsonLink`-related arguments here should be avoided.

comigor commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
-    };
-    request.headers.addAll(jsonHeaders);
-    return _httpClient.send(request);
+  Stream<exec.Response> request(exec.Request request,
```
Why is it a Stream? Would `Link` work with subscriptions?

klavs replied:
Yes, the intention is that `Link` works with any kind of operations.

comigor commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
-    request.headers.addAll(jsonHeaders);
-    return _httpClient.send(request);
+  Stream<exec.Response> request(exec.Request request,
+      [gql_link.NextLink forward]) {
```
Could you explain what `forward` means? (sorry about that, but `gql` has no docs on this)

klavs replied:
Forward calls the next `Link` in the link chain.

A `DedupeLink` would not call forward on a request that is already in flight.

klavs commented:
Sorry, I should give more explanations.

`Link` is modelled after Apollo Link. `Link` is basically a middleware. Which let's you implement features in a modullar fashion.

`Link` supports all types of GraphQL operations. It returns `Stream` because it has to support subscriptions.

GraphQL spec is intentionally protocol-agnostic. That's where `Link` comes in. There can be an `HttpLink`, `WebsocketLink`, etc. And it comes down to the user to chose a `Link` suiting their needs.

klavs commented:
It might appear that `Link` is taking away lot's of responsibilities from `ArtemisClient` but I think it's a good thing because Artemis is really about adding types around GraphQL requests while `Link` is all about executing GraphQL requests.

leoiacovini commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
+        ),
+      ),
+      data: jsonBody['data'],
+    );
```
According to the current spec of GraphQL `errors` may contain an `extensions` key with the additional data on `errors`: https://graphql.github.io/graphql-spec/June2018/#example-fce18 . And we should support.
Also, servers using an older version of the spec may still return extra keys in the same level as the `path`, `message` and `locations` - I think it would be good to not trim those, but we should at least follow the spec.

For me, this is a blocker given that we rely on extension data to handle errors properly

klavs replied:
That's a great point. I guess, response should also expose the raw response json.

klavs replied:
@leoiacovini on the second look, `extensions` on `error` are not yet implemented in Artemis. Which just point to the kind of problems that happen when every client implements it's own version of a common feature.

leoiacovini replied:
Sure thing. I think it would be great to have the `extensions` as a separated property (given that it's part of the spec) but should give access to the raw error map

klavs replied:
I am hesitant to add the raw "anything". Earlier I wrote "expose the raw response json", but today I realize that's protocol-specific and serialisation-format-specific.

I am now thinking that any non-standard behavior should be implemented by extending the `Response` to contain such non-standard fields. But it also means that the `Link` has to be extended. So that needs to be taken into account when implementing a more feature-ful `HttpLink`.

klavs replied:
I've now implemented the `HttpLink` in a way that any non-standard fields could be handled by extending base classes.

amadeu01 commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
-  /// The HTTP client which will be used to execute the request.
-  http.Client httpClient;
+  /// Instantiates an [ArtemisClient].
+  ArtemisClient({
```
Should it be an interesting addition having an named constructor `.fromUrl()` and then initialize a `SimpleHttpLink` on that, and break really little for other users?

klavs replied:
I'm thinking that `ArtemisClient` shouldn't rely on any specific `Link` since GraphQL and the generated code (by Artemis) is protocol-agnostic.

amadeu01 commented:
I like the concept of `Link`. I liked the implementation.
🎉

klavs commented on lib/client.dart:
```diff
@@ -1,59 +1,128 @@
+      graphQLEndpoint,
+      headers: {
+        'Content-type': 'application/json',
+        'Accept': 'application/json',
```
I'm not sure that's useful in Dart considering it's support for constants.

amadeu01 commented on lib/client.dart:
```diff
@@ -80,9 +84,18 @@ class ArtemisClient {
-  ArtemisClient({
-    this.link,
-  }) : assert(link != null);
+  factory ArtemisClient(
```
Nice usage of factory \o/

amadeu01 commented on example/graphbrainz/lib/main.dart:
```diff
@@ -6,9 +6,7 @@ import 'queries/ed_sheeran.query.dart';
-  final client = ArtemisClient(
-    link: SimpleHttpJsonLink(graphQLEndpoint),
-  );
+  final client = ArtemisClient(graphQLEndpoint);
```
Really nice !

leoiacovini commented on lib/client.dart:
```diff
@@ -1,59 +1,141 @@
+  }
+
+  Future<exec.Response> _post(exec.Request request) async {
+    final dataResponse = await _httpClient.post(
```
Instead of always calling the `POST` on the client WDYT about making this configurable (it can default to post)? I know that POST is the most usual method used, however, there are cases that use GET instead. (especially because we are trying to decouple the transmission protocol from the GraphQL itself with those changes)

amadeu01 replied:
I think this `SimpleHttpJsonLink` is just to easy for user that does not want to spend time implement their own Link.
But, I think it could be kept as simple as it is right now, and for more configurable one, the user should implement their own.
Also, I think it could be nice a little _how to_ about implement your own `Link`

leoiacovini replied:
I agree about it, however, I believe that the most "common" Links could be shipped with the package (just like Apollo does with Http, Auth, etc.. Links), this would eliminate the friction of implementing a whole new Link for just changing small settings (like Method, or Headers) and will make users able to get all goodies of the default implementation

klavs replied:
Yes, once Dart has the basic set of Links, user will mostly use those and extend them only when some rare special conditions are met.

Of course, we are not there yet. Even `package:graphql` implements only the must-have links.

klavs replied:
@leoiacovini I am all for supporting `GET` as well. But just today I read how `apollo-link-http` does it and it appears to be a bit too involved task to implement right now.

klavs commented:
@leoiacovini I'm definitely a fan of smaller more focused changes.

I think the scope of this PR should be to introduce `Link`. Any related nice-to-haves should better be a separate PR.

There's no rush on my side. My main goal was to start the conversation to see if there's agreement that ApolloLink-like interface would be useful and supported here as well.

comigor commented:
@klavs could you rebase and add some documentation to it before merging? I think maybe it's time to split readme or start using wiki, wdyt?

klavs commented:
I've moved the AST generation to build-time avoiding run-time parsing.

Also, I've implemented the `HttpLink` externally as a reusable `Link`. With that, the responsibility of serialising the AST and parsing the response is moved outside of Artemis. Transforming the `data` JSON into typed object is the responsibility of Artemis.

With this change `ArtemisClient` is a wrapper around `Link` adding Dart types on top of `data` JSON.

Note: I haven't reviewed the docs yet.

klavs commented:
@leoiacovini @amadeu01 @comigor

I consider this PR done. Feel free to accept it if you think it would benefit Artemis project.

I know it might appear to be a bit aggressive of me to come in and change a lot of code to packages I've implemented, but
- first, I genuinely believe a good separation of concerns benefits the whole Dart GraphQL ecosystem;
- do not consider `package:gql*` family of packages as mine, contributions are welcome.

This PR has been a great proof that the concepts in `package:gql` can work well with Artemis. I'm now moving to proof it in `package:graphql` client, where they have a similar concept of `Link`, but not as powerful because it is not based on AST. In the end `package:gql` will become better and more complete and, transitively, Artemis will become better.

Thanks for your feedback and suggestions.

comigor commented:
@klavs
I see value in having this separation of concerns as well, so Artemis could focus on the generation itself, from the pre-parsed AST (which needs some love, btw, but I didn't have time to do it yet).

Sadly, this PR is gonna be a breaking change due to moving `GraphQLError` to `gql`, but I don't think it makes sense to adapt to the `GraphQLError` we had here before.

Again, sorry for taking to long to review/merge/follow this, but I had a long week offsite.

Thanks again for all your help and PR's!

comigor commented:
Released as https://pub.dev/packages/artemis/versions/2.0.0
